src/dev.rs

```diff
@@ -407,24 +408,27 @@ impl<F: FieldExt> MockProver<F> {
 
         let mut cs = ConstraintSystem::default();
         let config = ConcreteCircuit::configure(&mut cs);
+        let cs = cs;
 
-        let fixed = vec![vec![None; n as usize]; cs.num_fixed_columns];
-        let advice = vec![vec![None; n as usize]; cs.num_advice_columns];
-        let permutations = cs
-            .permutations
-            .iter()
-            .map(|p| permutation::keygen::Assembly::new(n as usize, p))
-            .collect();
+        if n < cs.minimum_rows() {
+            return Err(Error::NotEnoughRowsAvailable);
+        }
+
+        let fixed = vec![vec![None; n]; cs.num_fixed_columns];
+        let advice = vec![vec![None; n]; cs.num_advice_columns];
+        let permutation = permutation::keygen::Assembly::new(n, &cs.permutation);
 
+        let blinding_factors = cs.blinding_factors();
         let mut prover = MockProver {
-            n,
+            n: n as u32,
             cs,
             regions: vec![],
             current_region: None,
             fixed,
             advice,
             instance,
-            permutations,
+            permutation,
+            usable_rows: ..n - (blinding_factors + 1),
```

Comment:
I think there's an off-by-one here. Let's say we need 5 blinding factors, and we set `n = 8`. Then this will be `..(8 - (5 + 1))` = `..2`, but after looking at `ConstraintSystem::minimum_rows` I'd expect the usable range to be `..1`. Perhaps this should be:
```suggestion
            usable_rows: ..n - unusable_rows,
```

Reply:
The off-by-one isn't here but in `minimum_rows()`, where I'm being strict (read: cautious) about leaving enough breathing room for the permutation argument. For example, I thought it would be weird if someone set up a circuit such that the permutation polynomial ends up looking like `[1, {0,1}, blind, ...]` over the evaluation domain, or even `[1, blind, ...]`. I didn't want to have to analyze these cases so I essentially force all circuits to have permutation polynomials that are at _least_ like `[1, something, {0,1}, blind, ...]` so there's clear separation for analysis purposes.

As an example, if `l_last = l_0` because we're not strict enough in this separation, then the proof could end up repeating the evaluations of the permutation polynomial in the proof. Could lead to bugs and it's weird.

I'll document this in `minimum_rows()` but I'm going to keep the existing (strict) behavior.

Reply:
In that case, document here that this `+ 1` is for `l_{-(m + 1)} (l_last)`, and that we do not need to prevent circuits from assigning to the `l_0` breathing-room row.

Reply:
```suggestion
            usable_rows: ..usable_rows,
```